src/storage/src/hummock/sstable/forward_sstable_iterator.rs

```diff
@@ -80,15 +137,77 @@ impl SstableIterator {
+                        tracing::warn!(target: "events::storage::sstable::block_seek", "prefetch mismatch: sstable_id = {}, block_id = {}, prefetched_block_id = {}",self.sst.value().id,idx,*prefetched_idx);
+                        self.prefetched_blocks.clear();
+                        true
+                    }
+                } else {
+                    true
+                };
+                let next_prefetch_idx = self
+                    .prefetched_blocks
+                    .back()
+                    .map_or(idx, |(latest_idx, _)| *latest_idx)
+                    + 1;
+                if is_empty && next_prefetch_idx > dest_idx {
+                    false
+                } else {
+                    let sst = self.sst.value();
+                    if is_empty {
+                        let (block_loc, uncompressed_capacity) = sst.calculate_block_info(idx);
+                        self.prefetched_blocks.push_back((
+                            idx,
+                            self.sstable_store
+                                .get_with_block_info(
+                                    sst.id,
+                                    idx as u64,
+                                    block_loc,
+                                    uncompressed_capacity,
+                                    crate::hummock::CachePolicy::Fill,
+                                    &mut self.stats,
+                                )
+                                .await?,
+                        ));
+                    }
+                    if next_prefetch_idx <= dest_idx {
+                        let (block_loc, uncompressed_capacity) =
+                            sst.calculate_block_info(next_prefetch_idx);
+                        self.prefetched_blocks.push_back((
+                            next_prefetch_idx,
+                            self.sstable_store
+                                .get_with_block_info(
+                                    sst.id,
+                                    next_prefetch_idx as u64,
+                                    block_loc,
+                                    uncompressed_capacity,
+                                    crate::hummock::CachePolicy::Fill,
+                                    &mut self.stats,
+                                )
+                                .await?,
+                        ));
+                    }
+                    true
+                }
+            } else {
+                false
+            };
+            let block = if in_prefetch {
+                self.prefetched_blocks.pop_front().unwrap().1.wait().await?
+            } else {
+                self.sstable_store
+                    .get(
+                        self.sst.value(),
+                        idx as u64,
+                        crate::hummock::CachePolicy::Fill,
+                        &mut self.stats,
+                    )
+                    .await?
+            };
```

Comment:
It is better to put the prefetch logic in a enum to improve the overall readability. Here is my suggestion:
```rust

enum BlockFetcher {
    Simple,
    Prefetch(PrefetchContext)
}

impl BlockFetcher {
    async fn get_block(
        &self,
        sst: &Sstable,
        block_idx: u64,
        sstable_store: SstableStoreRef
        stats: &mut StoreLocalStatistic,
    ) -> HummockResult<BlockHolder> {
        match self {
            BlockFetcher::Simple => {
                sstable_store
                .get(
                    sst,
                    block_idx,
                    crate::hummock::CachePolicy::Fill,
                    stats,
                )
                .await
            },
            BlockFetcher::Prefetch(context) => {
                context.get_block(sst, block_idx, sstable_store, stats).await
            },
        }
    }
}

struct PrefetchContext {
    prefetched_blocks: VecDeque<(usize, BlockResponse)>,

    /// block[cur_idx..=end_idx] will definitely be visited in the future.
    end_idx: usize,
}

impl PrefetchContext {
    pub fn new(start_idx: usize, end_idx: usize) -> Self {
        // Fetch the initial block at start_idx into prefetched_blocks
        ...

        Self {
            prefetched_blocks,
            end_idx
        }
    }
    async fn prefetch(&self, sstable_store: SstableStoreRef, stats: &mut StoreLocalStatistic) -> HummockResult<()> {
        let prefetch_idx = ...
        if prefetch_idx <= end_idx {
            let block_resp = sstable_store.async_get(...).await?
            self.prefetched_blocks.push_back((prefetch_idx, block_resp));
        }
        Ok(())
    }

    async fn get_block(
        &self,
        sst: &Sstable,
        block_idx: u64,
        sstable_store: SstableStoreRef
        stats: &mut StoreLocalStatistic,
    ) -> HummockResult<BlockHolder> {
        self.prefetch(...);
        let (idx, block_resp) = self.prefetched_blocks.pop_front().unwrap();
        assert_eq!(idx, block_idx);
        block_resp.wait().await
    }
}
```

We can store and initialize `BlockFetcher ` accordingly in SstableIterator based on the iter read option. Inside `seek_idx`, we just need to call `self.block_fetcher.get_block`. If we need to modify the pretch policy, we can easily do so by modifying `PrefetchContext::prefetch`

Reply:
> It is better to put the prefetch logic in a enum to improve the overall readability. Here is my suggestion:
> 
> ```rust
> enum BlockFetcher {
>     Simple,
>     Prefetch(PrefetchContext)
> }
> 
> impl BlockFetcher {
>     async fn get_block(
>         &self,
>         sst: &Sstable,
>         block_idx: u64,
>         sstable_store: SstableStoreRef
>         stats: &mut StoreLocalStatistic,
>     ) -> HummockResult<BlockHolder> {
>         match self {
>             BlockFetcher::Simple => {
>                 sstable_store
>                 .get(
>                     sst,
>                     block_idx,
>                     crate::hummock::CachePolicy::Fill,
>                     stats,
>                 )
>                 .await
>             },
>             BlockFetcher::Prefetch(context) => {
>                 context.get_block(sst, block_idx, sstable_store, stats).await
>             },
>         }
>     }
> }
> 
> struct PrefetchContext {
>     prefetched_blocks: VecDeque<(usize, BlockResponse)>,
> 
>     /// block[cur_idx..=end_idx] will definitely be visited in the future.
>     end_idx: usize,
> }
> 
> impl PrefetchContext {
>     pub fn new(start_idx: usize, end_idx: usize) -> Self {
>         // Fetch the initial block at start_idx into prefetched_blocks
>         ...
> 
>         Self {
>             prefetched_blocks,
>             end_idx
>         }
>     }
>     async fn prefetch(&self, sstable_store: SstableStoreRef, stats: &mut StoreLocalStatistic) -> HummockResult<()> {
>         let prefetch_idx = ...
>         if prefetch_idx <= end_idx {
>             let block_resp = sstable_store.async_get(...).await?
>             self.prefetched_blocks.push_back((prefetch_idx, block_resp));
>         }
>         Ok(())
>     }
> 
>     async fn get_block(
>         &self,
>         sst: &Sstable,
>         block_idx: u64,
>         sstable_store: SstableStoreRef
>         stats: &mut StoreLocalStatistic,
>     ) -> HummockResult<BlockHolder> {
>         self.prefetch(...);
>         let (idx, block_resp) = self.prefetched_blocks.pop_front().unwrap();
>         assert_eq!(idx, block_idx);
>         block_resp.wait().await
>     }
> }
> ```
> 
> We can store and initialize `BlockFetcher ` accordingly in SstableIterator based on the iter read option. Inside `seek_idx`, we just need to call `self.block_fetcher.get_block`. If we need to modify the pretch policy, we can easily do so by modifying `PrefetchContext::prefetch`

A slight difference is that when prefetch policy is on while no blocks need prefetching, this PR manages to call the original `get` interface but the modified version here will call `prefetch` interface. Though these two interfaces almost do the same thing at present, it will block further optimization.

Reply:
> put the prefetch logic in a enum to improve the overall readability

fixed